### PR TITLE
Added test for relations coherence

### PR DESCRIPTION
**Pull request by TakoB222**
### Issue
As desribed [here](https://github.com/canonical/postgresql-operator/issues/449), added test for checking relations coherence

### CODE REVIEW

File: tests/integration/new_relations/test_relations_coherence.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        await ops_test.model.applications[DATABASE_APP_NAME].remove_relation(
+            f"{DATABASE_APP_NAME}:database",
+            f"{APPLICATION_APP_NAME}:{FIRST_DATABASE_RELATION_NAME}",
+        )
+        await ops_test.model.wait_for_idle(apps=[DATABASE_APP_NAME], status="active", timeout=1000)
+
+        # Repeat re-relation using all available relation options (e.g. legacy_roles)
+        await ops_test.model.applications[APPLICATION_APP_NAME].set_config({
+            "legacy_roles": "true",
+        })
```
Comment (dragomirp) [inline tests/integration/new_relations/test_relations_coherence.py:108]: This won't work. The `legacy_roles` is there to check that the postgresql charm will block if those are requested and there is no way to satisfy that interface.

File: tests/integration/new_relations/test_relations_coherence.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        )
+
+        await ops_test.model.add_relation(
+            f"{APPLICATION_APP_NAME}:{FIRST_DATABASE_RELATION_NAME}", DATABASE_APP_NAME
+        )
+        await ops_test.model.wait_for_idle(apps=APP_NAMES, status="active")
+
+        primary_connection_string = await build_connection_string(
+            ops_test, APPLICATION_APP_NAME, FIRST_DATABASE_RELATION_NAME
+        )
```
Comment (dragomirp) [inline tests/integration/new_relations/test_relations_coherence.py:120]: Legacy interface on the test app is `db` not `first-database`

File: tests/integration/new_relations/test_relations_coherence.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        connection = psycopg2.connect(primary_connection_string)
+        connection.autocommit = True
+        cursor = connection.cursor()
+        random_name = f"test_{''.join(secrets.choice(string.ascii_lowercase) for _ in range(10))}"
+        cursor.execute(f"CREATE DATABASE {random_name};")
+        cursor.execute(f"DROP DATABASE {random_name};")
+        connection.close()
```
Comment (dragomirp) [inline tests/integration/new_relations/test_relations_coherence.py:127]: I don't think we can create DBs with the test app user. This should most probably be creating tables instead.

File: tests/integration/new_relations/test_relations_coherence.py (hunk 1 above)
Comment (marceloneppel) [reply tests/integration/new_relations/test_relations_coherence.py:108]: Alternative roles that can be used with the modern relation are, for example: `admin` (custom role created by the charm) and PostgreSQL predefined roles like `CREATEDB`, `CREATEROLE` and `SUPERUSER`.

Comment (taurus-forever) [reply tests/integration/new_relations/test_relations_coherence.py:108]: It was requested `extra-user-roles` [here](https://github.com/canonical/postgresql-operator/issues/449) to make sure switching from user->admin->user->admin works as expected. Admin has admin level and User has no admin level => test to make sure no garbage left in DB/databags.

File: tests/integration/new_relations/test_relations_coherence.py (hunk 3 above)
Comment (taurus-forever) [reply tests/integration/new_relations/test_relations_coherence.py:127]: +1 to @dragomirp , bit I am surprised to see [it passed](https://github.com/canonical/postgresql-operator/actions/runs/8812966437/job/24189993180#step:21:167).

Comment (marceloneppel) [reply tests/integration/new_relations/test_relations_coherence.py:127]: The test app has permissions for that: https://github.com/canonical/postgresql-test-app/blob/3edafc10b4b5aab29355820924b39c208bec76b1/src/charm.py#L35C1-L35C17

**Comment by taurus-forever** [comment]
This is a good start!

Unfortunately the test still doesn't check critical corner cases `User has no admin level`.
IMHO, create DB is necessary [here](https://github.com/canonical/postgresql-operator/pull/450/files#diff-ccc2438c49dfa6dc6b1bd9f9969cd3f3c1394a9450ecf933048d450fd4a26c2fR55) and [here](https://github.com/canonical/postgresql-operator/pull/450/files#diff-ccc2438c49dfa6dc6b1bd9f9969cd3f3c1394a9450ecf933048d450fd4a26c2fR122). In both cases it should be forbidden. The first case check that default user is not an admin... the second check makes sure previous admin flags successfully removed and relation degraded to non-admin level.

Also I would try to connect as non-admin/admin another schemas to make sure they are reachable (e.g. `postgres`) additionally to newly created one `{random_name}`. Also I would not frop `{random_name}` and will ensure no access after re-relating as non-admin.

The test is a good start, but not enough to sleep well.

**Review by marceloneppel — APPROVED** [review]

**Review by marceloneppel — CHANGES_REQUESTED** [review]
I'm seeing some errors in the test now, related to schema change: https://github.com/canonical/postgresql-operator/actions/runs/9390124988/job/25898213829?pr=450#step:23:340

**Comment by dragomirp** [comment]
> I'm seeing some errors in the test now, related to schema change: https://github.com/canonical/postgresql-operator/actions/runs/9390124988/job/25898213829?pr=450#step:23:340

Schema change should be database change. I have confirmed with @taurus-forever

**Comment by taurus-forever** [comment]
Apologies for schema/database typo above. Rush.
P.S. non-admin uses have access to all schemas of his database. So we are checking: "no access to foreign database after devoting relation from admin to non-admin".

P.S. link is not happy: https://github.com/canonical/postgresql-operator/actions/runs/9413308332/job/25929799907?pr=450

File: tests/integration/new_relations/test_relations_coherence.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        # Re-relation with admin role with checking permission
+        await ops_test.model.applications[DATA_INTEGRATOR_APP_NAME].set_config({
+            "database-name": DATA_INTEGRATOR_APP_NAME.replace("-", "_"),
+            "extra-user-roles": "admin",
+        })
+        await ops_test.model.wait_for_idle(apps=[DATA_INTEGRATOR_APP_NAME], status="blocked")
+        await ops_test.model.add_relation(DATA_INTEGRATOR_APP_NAME, DATABASE_APP_NAME)
+        await ops_test.model.wait_for_idle(apps=APP_NAMES, status="active")
+
+        connection_string = await build_connection_string(
```
Comment (taurus-forever) [inline tests/integration/new_relations/test_relations_coherence.py:93]: @marceloneppel please consider to assign https://warthogs.atlassian.net/browse/DPE-2278 to Lucas the next pulse. DPE lib supports it for Mongo already, we should expand it for PostgreSQL (and MySQL?) to simplify our life here and users copy&paste life as well :-)

**Review by taurus-forever — APPROVED** [review]
The code/test itself: LGTM, Thank you!
Please fix/pass CI tests.

**Comment by marceloneppel** [comment]
Hi, @TakoB222! The upgrade integration tests will fail until https://github.com/canonical/postgresql-operator/pull/492 is merged to the main branch and later merged to this PR branch.

**Comment by marceloneppel** [comment]
> > Hi, @TakoB222! The upgrade integration tests will fail until #492 is merged to the main branch and later merged to this PR branch.
> 
> Okay, thanks!

That PR is merged now. You can merge the `main` branch into yours.

**Comment by TakoB222** [comment]
> > > Hi, @TakoB222! The upgrade integration tests will fail until #492 is merged to the main branch and later merged to this PR branch.
> > 
> > 
> > Okay, thanks!
> 
> That PR is merged now. You can merge the `main` branch into yours.

Done

**Review by dragomirp — APPROVED** [review]

**Review by marceloneppel — APPROVED** [review]
